pkg/mauth/compare/compare.go

```diff
@@ -0,0 +1,311 @@
+// Package compare provides various compare operators with binary codecs
+// useful for macaroon-based authorization.
+package compare
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"regexp"
+)
+
+type Type uint8
+
+const (
+	TypeFalse    Type = 0
+	TypeTrue     Type = 1
+	TypeIntegers Type = 2
+	TypeStrings  Type = 3
+	TypeRegexp   Type = 4
+	TypeCIDRs    Type = 5
+)
+
+func UnmarshalBinary(b []byte) (interface{}, error) {
+	if len(b) < 1 {
+		return nil, fmt.Errorf("compare: buffer to unmarshal is zero-length")
+	}
+	switch Type(b[0]) {
+	case TypeFalse, TypeTrue:
+		v := new(Bool)
+		return *v, v.UnmarshalBinary(b)
+	case TypeIntegers:
+		v := new(Integers)
+		return *v, v.UnmarshalBinary(b)
+	case TypeStrings:
+		v := new(Strings)
+		return *v, v.UnmarshalBinary(b)
+	case TypeRegexp:
+		v := new(Regexp)
+		return v, v.UnmarshalBinary(b)
+	case TypeCIDRs:
+		v := new(CIDRs)
+		return *v, v.UnmarshalBinary(b)
+	default:
+		return nil, fmt.Errorf("compare: unknown type %d", b[0])
+	}
+}
+
+type Bool bool
+
+func (want Bool) Compare(have bool) bool {
```

Comment:
Perhaps `s/want/b`, not a material change but I had a double take when scanning function signatures because of the inconsistency.